### PR TITLE
Design doc for access control management UI

### PR DESCRIPTION
**Pull request by mlamina**
WIP, trying to formalize the ideas from #1441.

### CODE REVIEW

**Comment by bryk** [comment]
This is a good start @mlamina! I see that this is a template only as of now, which is fine. Should we marge it or wait for you to fill the contents? Both ways work for me.


In the meantime, try to fix the issue with the CLA. Click "details" link on the red line above :)

**Comment by bryk** [comment]
Btw, join SIG-UI meeting tomorrow to talk about this, if you can. 

Details here: https://github.com/kubernetes/community/tree/master/sig-ui

**Comment by mlamina** [comment]
I signed the CLA!
I would suggest to keep the doc in this PR right now to have an active discussion?

**Comment by mlamina** [comment]
Aright, I worked on the design document to make it a little more complete. Would be great to get some feedback from @danielromlein :)

**Comment by cheld** [comment]
CC @floreks, @kenan435

**Comment by danielromlein** [comment]
Questions / thoughts after looking through the [design doc](https://github.com/mlamina/dashboard/blob/33439145126fca1ab03584d358219c7e05206d16/docs/design/access-control.md):
- I think it'd be great to flesh out the use cases more. _Why_ as a cluster admin do I want to create roles and assign them to users? 
- What sort of ranges do we expect to see for Roles and Bindings? If they're going to potentially grow large, might it make sense to use tabs for them, or to paginate the table?
- Do we want some view where a user can see both what a role’s permissions are _and_ to whom that roll is assigned? Currently only link between them.
- I'm struggling to understand Bindings... Would there ever be a reason to have multiple bindings for the same role in the same namespace?
- @bryk Would this require further engineering work to get implemented, or is Dashboard set up to handle RBAC today? I'm just curious.

@normalnorman your input around this would be super valuable, as you have more experience with PRDs than me 😊

**Comment by bryk** [comment]
Sorry @mlamina if I'm stealing your answers here :) Feel free to correct me. 

> What sort of ranges do we expect to see for Roles and Bindings? If they're going to potentially grow large, might it make sense to use tabs for them, or to paginate the table?

Dozens of roles and bindings is what I'd expect for a large org. I think pagination or a way of handling large number of roles is a must here. 

>@bryk Would this require further engineering work to get implemented, or is Dashboard set up to handle RBAC today? I'm just curious.

We can handle RBAC admin as it is here easily. If a tricky role is then assigned to the UI pod, it'll break. In Q1 it should be handling all roles.

File: docs/design/access-control.md [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+
+## Abstraction / Simplification
+
+The only difference between Roles and ClusterRoles is that Roles apply to a specific namespace 
+and ClusterRoles apply to all namespaces. The same is true for RoleBindings and ClusterRoleBindings. 
+The UI should help users to understand these concepts and user flows should be designed to enforce
+the correct usage. 
+
+NOTE: The design proposed in this document will combine certain concepts:
```
Comment (bryk) [inline docs/design/access-control.md]: Interesting proposal. I think I like it. Simplifies the problem from user's perspective. We could say in some non-prominent place that * means that this is cluster role object, but overall this looks nice.

WDYT? @rf232 @normalnorman

File: docs/design/access-control.md [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+* Edit a role's rules
+* Unbind a subject from a role
+
+# Design
+
+
+
+## Overview
+
+The overview can be reached by clicking the "Access Control" item in the admin section of the
```
Comment (bryk) [inline docs/design/access-control.md]: Wondering about name here. Should this be called "Access Control"? Or maybe "RBAC"? "Roles and Bindings"? Maybe "IAM"? Or just "Roles" as bindings are auxiliary objects? 

Listing a couple of ideas to explore the space.

@rf232 @normalnorman what do you think?

File: docs/design/access-control.md [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+left-hand menu. It shows existing Roles/Bindings, some important details for every item and allows
+the user to create a new Role or Binding right away.
+
+![Overview](mockups/21-11-2016-access-control/overview.png)
```
Comment (bryk) [inline docs/design/access-control.md]: Mock LGTM

File: docs/design/access-control.md [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+## Creating a role
+
+Clicking the "+" under the existing Roles will open a dialog. After entering the name and
```
Comment (bryk) [inline docs/design/access-control.md]: This is a tricky part. We've never had such proposal for creating objects. All we had is this +create button, which is mildly useful:
image

I like the proposal here to move creation to the context of where it is used. We can apply this pattern project wide if it turns out to be good.


Other option would be to add "+create" button to the list toolbar, as this is our long term place to put things related to a list displayed below.  

@normalnorman @danielromlein What do you think re this?

File: docs/design/access-control.md [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+picking a namespace ("*" will create a ClusterRole), clicking "Create" will create the Role
```
Comment (bryk) [inline docs/design/access-control.md]: I'm wondering whether the `*` is too magic. From my talks with users I remember that they do not like magic.

How about changing `*` to `all namespaces` or `all namespaces (will create a ClusterRole)` or add `info` bubble somewhere there to explain this. All I want here is to be explicit to what we do.

File: docs/design/access-control.md [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+
+Clicking the "+" under the existing Bindings will open a dialog. After the selecting the
+target role and the namespace ("*" will create a ClusterRoleBinding), clicking "Create"
+will create the Binding and redirect the user to the Binding's detail page.
+
+![Overview](mockups/21-11-2016-access-control/create-binding.png)
+
+## Edit a role
+
+The only property of a Role that cannot be changed is its namespace, because it determines
```
Comment (bryk) [inline docs/design/access-control.md]: How do I get into the state of editing a role? What do I click?

File: docs/design/access-control.md [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+
+## Bind subjects to a role
+
+Clicking the "+" under the existing Bindings will open a dialog. After the selecting the
+target role and the namespace ("*" will create a ClusterRoleBinding), clicking "Create"
+will create the Binding and redirect the user to the Binding's detail page.
+
+![Overview](mockups/21-11-2016-access-control/create-binding.png)
+
+## Edit a role
```
Comment (bryk) [inline docs/design/access-control.md]: Do you propose a page that would show details of a role? E.g., the pages that we currently have for other objects.

File: docs/design/access-control.md (hunk 5 above)
Comment (mlamina) [inline docs/design/access-control.md]: Yes, I think `all namespaces` would be more explicit and having an info bubble for further clarification is a good idea!

File: docs/design/access-control.md (hunk 6 above)
Comment (mlamina) [inline docs/design/access-control.md]: My idea was to have clickable list items in the overview. Clicking an item leads to the edit page.

File: docs/design/access-control.md (hunk 7 above)
Comment (bryk) [inline docs/design/access-control.md]: I would kind of expect a page to show single binding or single role. Just to be parallel with all other views in the UI. But I see your point that such view is not required.

I guess this is call to action for out friendly UX designers who should have opinions on this. 

@danielromlein @normalnorman Can you suggest something here?

File: docs/design/access-control.md (hunk 6 above)
Comment (bryk) [inline docs/design/access-control.md]: Typically what we do in other views is that when we click on a resource, we show its details page. Then there user can edit the thing. Does this sound sane?

Comment (danielromlein) [inline docs/design/access-control.md]: To me it makes sense to have a detail page with an edit option there. RE my earlier question
>Do we want some view where a user can see both what a Role’s permissions are _and_ to whom that Role is assigned?

The detail page would give us the opportunity to display information like, for example, what Bindings a given Role applies to (if that's important to users...?)

Comment (mlamina) [inline docs/design/access-control.md]: Good idea! Would anyone mind to make some changes in the mockups? I joined a new project at work and am quite busy at the moment, but I'll do my best to keep the ball rolling!

Comment (danielromlein) [inline docs/design/access-control.md]: @mlamina I'll try to make changes! Right now my priority is the Pod Details page #1248, but this is next on my list.

**Comment by mlamina** [comment]
@bryk @danielromlein @normalnorman I updated the mockups and the proposal. There is now a separate detail page for roles and instead of the "*", we use "all namespaces". I thought about adding a separate detail page for Roles as well, but there wouldn't be any difference between the detail page and the edit page, because all the user would be able to do is add or remove subjects. Ideas? Comments?

File: docs/design/access-control.md (hunk 2 above)
Comment (mlamina) [inline docs/design/access-control.md]: IMHO "Access Control" is the most precise and unambiguous term?

**Comment by danielromlein** [comment]
One Role can have multiple Bindings, correct? I'm not seeing how that would work in the current [wireframe](https://github.com/mlamina/dashboard/blob/7fe272e1bef40d7cd1f3b3df146a289d24d9a428/docs/design/access-control.md#view-a-role). Is there a wireframe accommodating that?

I think we'd agreed putting the _Create_ buttons up [at the top of each card](https://invis.io/969TJXX4B#/208427844_Wireframe) was a better solution and more in-line with the Material spec than below.

👍  on the change to "All Namespaces."

**Comment by mlamina** [comment]
@danielromlein yes, a Role can have multiple Bindings. The "Role detail view" you referenced even shows it: The "details" section shows the Role's properties while the "Bindings" section below that lists the Bindings associated with that Role.

You're absolutely right, the Create buttons should be at the top of each card, I'll make that adjustment. Great mockups by the way! :)

**Comment by bryk** [comment]
Lets merge! @mlamina let us know when you need any help coding this.

**Comment by mlamina** [comment]
Thanks @bryk ! Should I just create another pull request and get on with it or are there any procedures / best practices that you guys would like me to follow?

**Comment by bryk** [comment]
Just create a pull request :)

**Comment by danielromlein** [comment]
Sorry I missed this @mlamina! Yeah go ahead and create a PR this LGTM.

**Comment by mlamina** [comment]
https://github.com/kubernetes/dashboard/pull/1576

**Comment by mlamina** [comment]
OK so I started implementing and found that the design is missing an important detail: I assumed that a role simply consists of a set of verbs and a set of resources. However, A role can have multiple rules and each of these rules contains one set of resources and one set of verbs. This clearly not only impacts the design, but also the implementation. Now I see two possible ways to handle this:

1. Adjust the design to account for the 1:n relationship between roles and rules
2. Restrict the implementation to a 1:1 relationship

Option (1) is obviously the easy way out and might need to be fixed in the future. However, option (2) would probably have a heavy impact on the design document. Would love to hear your opinions!

**Comment by bryk** [comment]
I'd suggest adjusting the design. We should create UIs that handle most cases. If we handle only 1:1 relationships this would be unusable for users who have 1:n relationships.

How much does this affect the implementation? Can you comment?

**Comment by mlamina** [comment]
Yes, I agree! The backend part of the implementation should not be affected too much, it's just a small change in data types. However the fronted would look very different, since every role can now have a list of "rules" instead of just some verbs and resources. I'm not a designer, so if somebody with more expertise in this could take a look at it that would be cool! :)

**Comment by danielromlein** [comment]
Interactive [wireframes](https://invis.io/2GA8MURUF) incorporating the concept of 1:n Rules per Role. 
@mlamina PTAL
I still don't have a good understanding of API groups.
>look at them simply as a list of strings

☝️ I don't know what you mean by this...?

**Comment by mlamina** [comment]
OK maybe check out [this link](https://kubernetes.io/docs/api/) under "API Groups"? They are a concept to split API endpoints into groups, which makes them easier to upgrade and manage.
In the context of Roles and their rules, this is how you can think of the semantics of a rule:

> Allow all subjects bound to this rule to apply [list of verbs] to [list of resources] under [list of API groups]

[list of X] is simply a list of strings, so for example we could have something like:

apiGroups: ["", "rbac.authorization.k8s.io", "some.other.group"]
resources: ["pods"]
verbs: ["get", "watch", "list"]

**Comment by danielromlein** [comment]
Ok @mlamina that makes sense – thanks for the explanation!
WDYT about the [wireframes](https://invis.io/7Z9NNP4E8)?
cc @normalnorman thoughts on this IA pattern?
/ @bryk

**Comment by mlamina** [comment]
Oh right, forgot my feedback on the wireframes :)
- I love the split pane design! 
- When creating a role, it might be useful to have a "done" or "create" button to finish the action
- The popup for creating a rule is missing a field for API groups
- What does the "+ Create" button in the upper right corner do?
- In the popup for creating a RoleBinding:
  - the first dropdown is probably supposed to say "Role" for selecting a Role, not "Rule"?
  - the "Role Type" can be inferred from the selected Role, so we could simplify the view by removing this field
  - Is  the idea to only have one subject to begin with or to be able to add multiple subjects right away? Just wondering because there is no "+ Subject" button

**Comment by danielromlein** [comment]
@mlamina thanks! 👍 
 > When creating a role, it might be useful to have a "done" or "create" button to finish the action

There's Create / Cancel, but your missing them might indicate that they should be closer in proximity to inputs 😊 
 > The popup for creating a rule is missing a field for API groups

Should this be a multi-select pattern?
 > What does the "+ Create" button in the upper right corner do?

That's our global "Create" that's present across all pages that brings a user to the _Deploy a Containerized App_ page. We haven't tested that very methodically though to assess if people get confused with page-specific actions.

> the first dropdown is probably supposed to say "Role" for selecting a Role, not "Rule"?

Correct. Changed.
> the "Role Type" can be inferred from the selected Role, so we could simplify the view by removing this field

Great. Changed.
> Is the idea to only have one subject to begin with or to be able to add multiple subjects right away? Just wondering because there is no "+ Subject" button

Yes. My thought was once a user had entered values for both inputs, another set of inputs would appear below. Think that's clear enough?

I should have recommended this, but for future feedback it's easier to track in InVision itself (toggle on "Comments" in the lower RH corner, then click anywhere to leave a comment)

Where I'd like to go next with this is to mock it up in Sketch / InVision so we have a clickable prototype that we could test with Kubernetes users.

**Comment by mlamina** [comment]
> Should this be a multi-select pattern?

I'm not exactly sure what the multi-select pattern looks like, but the end result needs to be a list of (possibly arbitrary) strings!

> That's our global "Create" that's present across all pages that brings a user to the Deploy a Containerized App page

For me personally it was a bit confusing to see that button in the _Access Control_ section. I've always wondered: Are there usage statistics / metrics being collected in the dashboard? Do we know which parts of the UI are used and in which way?

> My thought was once a user had entered values for both inputs, another set of inputs would appear below

Cool! I like it :)

> I should have recommended this, but for future feedback it's easier to track in InVision itself

Good idea, will do that next time! Looking forward to seeing the clickable prototype!

**Comment by danielromlein** [comment]
>Are there usage statistics / metrics being collected in the dashboard?

There are not, but I agree that this would be tremendously valuable to UX. We opened #1423 to explore the subject of telemetry but it hasn't gone anywhere. If you're willing to drive that forward I'd encourage you to go for it!

**Comment by danielromlein** [comment]
Interactive [InVision prototype](https://invis.io/9HAB8JPMW) I'd like to get people's feedback on. Click around and let me know what you think of the flow. Are any use cases not accounted for?

Please comment using the built-in [InVision functionality](https://support.invisionapp.com/hc/en-us/articles/209192426-How-do-I-comment-on-a-prototype-) so we can keep track of what comments apply to which areas of the UI.

cc @mlamina @bryk @rf232

**Comment by bryk** [comment]
I looked at InVision mocks. They look very solid! I left a few comments.

**Pull request merged**